File: src/aoc2024/day1.rs
use std::collections::HashMap;
use std::io::BufRead;

fn part_a(lines: &Vec<String>) {
    let values: Vec<Vec<i32>> = lines
        .iter()
        .map(|x| {
            x.split_whitespace()
                .map(|y| y.parse::<i32>().unwrap())
                .collect()
        })
        .collect();

    let mut left: Vec<i32> = values.iter().map(|x| *x.first().unwrap()).collect();
    let mut right: Vec<i32> = values.iter().map(|x| *x.last().unwrap()).collect();

    left.sort();
    right.sort();

    let result: i32 = left
        .into_iter()
        .zip(right.into_iter())
        .map(|(a, b)| i32::abs(a - b))
        .sum();

    println!("{result}")
}

fn part_b(lines: &Vec<String>) {
    let values: Vec<Vec<i32>> = lines
        .iter()
        .map(|x| {
            x.split_whitespace()
                .map(|y| y.parse::<i32>().unwrap())
                .collect()
        })
        .collect();

    let left: Vec<i32> = values.iter().map(|x| *x.first().unwrap()).collect();
    let mut right: HashMap<i32, i32> = HashMap::new();

    values.iter().for_each(|x| {
        let key = x.last().unwrap();
        match right.get(key) {
            Some(count) => {
                right.insert(*key, count + 1);
            }
            None => {
                right.insert(*key, 1);
            }
        }
    });

    let result: i32 = left
        .into_iter()
        .map(|x| match right.get(&x) {
            Some(i) => x * i,
            None => 0,
        })
        .sum();

    println!("{result}")
}

pub fn run() {
    let lines: Vec<String> = std::io::stdin()
        .lock()
        .lines()
        .map(|x| x.unwrap())
        .collect();

    part_a(&lines);
    part_b(&lines);
}

File: python/src/aoc2024/day1.py
import sys


def part_a(lines: list[str]):
    left = []
    right = []

    for line in lines:
        s = line.strip().split()
        left.append(int(s[0]))
        right.append(int(s[1]))

    left.sort()
    right.sort()

    result = 0

    for a, b in zip(left, right):
        result += abs(a - b)

    print(result)


def part_b(lines: list[str]):
    left = []
    right = {}

    for line in lines:
        s = line.strip().split()
        left.append(int(s[0]))

        r = int(s[1])
        right[r] = right.get(r, 0) + 1

    result = 0
    for l in left:
        result += l * right.get(l, 0)

    print(result)


def run():
    lines = [line for line in sys.stdin]
    part_a(lines)
    part_b(lines)


if __name__ == "__main__":
    run()
